src/spectrum/group.rs

```diff
@@ -692,47 +693,43 @@ mod mzsignal_impl {
         }
     }
 
-    /// Average a series of [`MultiLayerSpectrum`] together
-    pub fn average_spectra<C: CentroidLike + Default, D: DeconvolutedCentroidLike + Default>(
-        spectra: &[MultiLayerSpectrum<C, D>],
+    /// Average a series of [`SpectrumLike`] together. The supplied dx will be used to [`reprofile`]
+    /// centroid spectra. The resulting [`ArrayPair`] can be made into a [`BinaryArrayMap`] using
+    /// `.into()`. Which in turn can be used to create a new [`MultiLayerSpectrum`] or
+    /// [`RawSpectrum`](crate::spectrum::RawSpectrum). Internally it uses [`SpectrumLike::peaks`]
+    /// to retrieve the spectrum data. Any deconvoluted spectra will be skipped.
+    ///
+    /// Note: only available with feature `mzsignal`.
+    pub fn average_spectra<
+        'lifetime,
+        S: SpectrumLike<C, D> + 'lifetime,
+        C: CentroidLike,
+        D: DeconvolutedCentroidLike,
+    >(
+        spectra: impl IntoIterator<Item = &'lifetime S>,
         dx: f64,
-    ) -> ArrayPair<'_> {
-        let arrays: Vec<_> = spectra
-            .iter()
-            .map(|scan| {
-                if scan.signal_continuity() == SignalContinuity::Profile {
-                    if let Some(array_map) = scan.raw_arrays() {
-                        let mz = array_map.mzs().unwrap().to_vec();
-                        let inten = array_map.intensities().unwrap().to_vec();
-                        Some(ArrayPair::from((mz, inten)))
-                    } else {
-                        warn!("{} did not have raw data arrays", scan.id());
-                        None
+    ) -> ArrayPair<'static> {
+        average_signal(
+            &spectra
+                .into_iter()
+                .flat_map(|spectrum| match spectrum.peaks() {
+                    RefPeakDataLevel::Missing | RefPeakDataLevel::Deconvoluted(_) => None,
+                    RefPeakDataLevel::RawData(array_map) => {
+                        let mzs = array_map.mzs().unwrap().to_vec();
+                        let intensities = array_map.intensities().unwrap().to_vec();
+                        Some(ArrayPair::from((mzs, intensities)))
```

Comment:
This still needs to check if the spectrum is centroid or profile. A centroid spectrum may be present in a raw state as unpacked arrays but not yet rolled into peak data structures. If you have arrays of centroids, the points will need to be re-profiled just as when you deal with the `Centroid` branch.

Reply:
Good point I missed that case indeed.